core/src/main/scala/org/apache/spark/util/CheckpointingIterator.scala

```diff
@@ -0,0 +1,225 @@
+    context: TaskContext,
+    blockSize: Int) extends Iterator[T] with Logging {
+
+  private[this] var completed = false
+
+  context.addTaskCompletionListener { ctx =>
+    // We don't know if the task is successful. So it's possible that we still checkpoint the
+    // remaining values even if the task is failed.
+    // TODO optimize the failure case if we can know the task status
+    complete()
+  }
+
+  private[this] val fileOutputStream = {
+    val bufferSize = SparkEnv.get.conf.getInt("spark.buffer.size", 65536)
+    if (blockSize < 0) {
+      fs.create(tempOutputPath, false, bufferSize)
+    } else {
+      // This is mainly for testing purpose
+      fs.create(tempOutputPath, false, bufferSize, fs.getDefaultReplication, blockSize)
+    }
+  }
+
+  private[this] val serializeStream =
+    SparkEnv.get.serializer.newInstance().serializeStream(fileOutputStream)
+
+  /**
+   * Called when this iterator is on the latest element by `hasNext`.
+   * This method will rename temporary output path to final output path of checkpoint data.
+   */
+  private[this] def complete(): Unit = {
+    if (completed) {
+      return
+    }
+
+    if (serializeStream == null) {
+      // There is some exception when creating serializeStream, we only need to clean up the
+      // resources.
+      cleanup()
+      return
+    }
+
+    while (values.hasNext) {
+      serializeStream.writeObject(values.next)
+    }
+
+    completed = true
+    CheckpointingIterator.releaseLockForPartition(id)
+    serializeStream.close()
+
+    if (!fs.rename(tempOutputPath, finalOutputPath)) {
+      if (!fs.exists(finalOutputPath)) {
+        logInfo("Deleting tempOutputPath " + tempOutputPath)
+        fs.delete(tempOutputPath, false)
+        throw new IOException("Checkpoint failed: failed to save output of task: "
+          + context.attemptNumber + " and final output path does not exist")
+      } else {
+        // Some other copy of this task must've finished before us and renamed it
+        logInfo("Final output path " + finalOutputPath + " already exists; not overwriting it")
+        fs.delete(tempOutputPath, false)
+      }
+    }
+  }
+
+  private[this] def cleanup(): Unit = {
+    completed = true
+    CheckpointingIterator.releaseLockForPartition(id)
+    if (serializeStream != null) {
+      serializeStream.close()
+    }
+    fs.delete(tempOutputPath, false)
+  }
+
+  override def hasNext: Boolean = {
+    try {
+      val r = values.hasNext
+      if (!r) {
+        complete()
+      }
+      r
+    } catch {
+      case e: Throwable =>
+        try {
+          cleanup()
+        } catch {
+          case NonFatal(e1) =>
+            // Log `e1` since we should not override `e`
+            logError(e1.getMessage, e1)
+        }
+        throw e
+    }
+  }
+
+  override def next(): T = {
+    try {
+      val value = values.next()
+      serializeStream.writeObject(value)
+      value
+    } catch {
+      case e: Throwable =>
```

Comment:
maybe we should put this in a private method? Something like `def cleanupOnFailure(body: => T)`
```
try {
  body
} catch {
  case NonFatal(e) => ...
}
```